### ISSUE TITLE
make install copies libfunctionality again (affects linux packaging)

### ISSUE THREAD

**Issue by ExpertSole** [OSVR/OSVR-Core#244]
I hacked together a couple of Gentoo ebuilds for installing libfunctionality and osvr-core with the Portage package manager. In testing them, I encountered what is ostensibly a bug in your cmake files; libfunctionality, which must have already been installed prior, is copied to /usr/{lib,lib64} along with osvr-core's files. This subtlety is one that Linux package maintainers will trip over briefly, because some package managers (Portage, for example) are strict about file collisions and will refuse to install OSVR if the install image collides with files belonging to another program.

steps to reproduce:
   cmake ./; make; make install
   (or use an ebuild to do it for you...)
actual result:
   all executable files/libraries referenced in the cmake files, including libfunctionality, are copied
expected result:
   only new (osvr-core) executable files/libraries should be copied

I was able to install osvr-core using ebuilds once I worked around the file collision. I don't have an HDK, so I haven't verified that the sample program works, but everything else checks out.

**Comment by stephenmsnow** [comment]
I have an HDK, and I'd be happy to help troubleshoot if you'd like to work
with me and provide repro/troubleshooting steps.  I'm also a programmer, so
I'd be happy to work with you to fix any issues (though my biggest strength
is also my weakness... I'm a C programmer, I am decent in C++, but I never
really developed a love for it, like so many linux people).  Let me know
what I can do to help, I can also give you my cell # if you think that'd
expedite the process.

**Comment by stephenmsnow** [comment]
I'm not running gentoo, I run debian and ubuntu, and for servers I run
CentOS, which I have X access to, but if I need to install gentoo, I'd be
more than happy to to help my fellow linux peeps :)

**Comment by stephenmsnow** [comment]
BTW... this isn't a case of TL;DR, it's a case of not knowing if your issue
is expected to be Gentoo specific or not.  I've always been a
Slackware/Debian guy... using RHEL/CentOS for servers, I've used just about
every distro as my primary in the last 15 years BUT Gentoo, so forgive me
if you already gave me the steps necessary to repro, and it was
linux-specific and not Gentoo-specific.

**Comment by stephenmsnow** [comment]
Also, I run CLion and Code::Blocks, if you'd like me to use another IDE to
keep things consistent with your work, let me know and I'll get it done.

**Comment by ExpertSole** [comment]
It's not really a Gentoo-specific problem, Gentoo is just the first distro I encountered it on. Any distro with a strict enough package manager will require a similar fix, so it might as well be fixed upstream. The problem is that make install installs files that belong to another program (libfunctionality); That's a no-go because it will require a workaround in packaging later. Someone who knows CMake has to remove a reference to libfunctionality from OSVR's CMake files. If that reference is necessary on Windows, maybe it should be conditional on a platform check?

I think this is the offending code, but I will defer to the CMake gurus if it's not. The last line of ./src/osvr/PluginHost/CMakeLists.txt reads:
"osvr_copy_dep(${LIBNAME_FULL} libfunctionality::functionality)"

TL;DR, on Linux, if libfunctionality is already installed, don't try to install it again, or hilarity will ensue. Fix CMake.

**Comment by rpavlik** [comment]
Yeah - does this make sense to have behind an option? I was anticipating distros packaging this eventually, but expected at first people to be installing to /opt and thus needing something self-contained. (Have also been considering making libfunctionality a submodule).  We're already doing some conditional installs for Windows (like OpenCV binaries, for instance), libfunc would be easy to do the same to.

**Comment by rpavlik** [comment]
and yes, you found the right line.

**Comment by d235j** [comment]
What's odd is that I'm not experiencing this on OS X.

**Comment by ExpertSole** [comment]
@d235j I just re-ran my ebuild without the temporary fix I applied before, and it failed. That process also executes git pull, so I believe that means the problem is not fixed in the current version on git upstream.

localhost ~ # emerge osvr-core
... (much spam related to configuration and compilation skipped)
>>> Failed to install dev-games/osvr-core-9999, Log file:

>>>  '/tmp/portage/dev-games/osvr-core-9999/temp/build.log'

 * Messages for package dev-games/osvr-core-9999:

 * This package will overwrite one or more files that may belong to other
 * packages (see list below). You can use a command such as `portageq
 * owners / <filename>` to identify the installed package that owns a
 * file. If portageq reports that only one package owns a file then do
 * NOT file a bug report. A bug report is only useful if it identifies at
 * least two or more packages that are known to install the same file(s).
 * If a collision occurs and you can not explain where the file came from
 * then you should simply ignore the collision since there is not enough
 * information to determine if a real problem exists. Please do NOT file
 * a bug report at http://bugs.gentoo.org unless you report exactly which
 * two packages install the same file(s). See
 * http://wiki.gentoo.org/wiki/Knowledge_Base:Blockers for tips on how to
 * solve the problem. And once again, please do NOT file a bug report
 * unless you have completely understood the above message.
 * 
 * Detected file collision(s):
 * 
 * 	/usr/lib64/libfunctionality.so.0.1
 * 
 * Searching all installed packages for file collisions...
 * 
 * Press Ctrl-C to Stop
 * 
 * dev-util/libfunctionality-9999:0::x-portage
 * 	/usr/lib64/libfunctionality.so.0.1
 * 
 * Package 'dev-games/osvr-core-9999' NOT merged due to file collisions.
 * If necessary, refer to your elog messages for the whole content of the
 * above message.

 * GNU info directory index is up-to-date.
localhost ~ # cd /usr/distfiles/git3-src/OSVR_OSVR-Core.git/
localhost OSVR_OSVR-Core.git # ls -lct |head -n 2 
total 32
-rw-r--r--  1 portage portage   23 Nov  1 23:19 HEAD

**Comment by stephenmsnow** [comment]
Hi,

I'd like to add you to my professional network on LinkedIn.

- Stephen

Confirm that you know Stephen: https://www.linkedin.com/e/v2?e=8j7g4x-igj9clvr-4o&t=ssuw&tracking=eml-guest-invite-cta&ek=invite_guest&sharedKey=qBOkYmfz&invitationId=6067262158750699520








You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=8j7g4x-igj9clvr-4o&t=lun&midToken=AQFmrFVhGEd36Q&ek=invite_guest&loid=AQGkbbOq_2QS_wAAAVDM-6qzRKPu6TtQ7Zf3-l_F-Lmlxuq_Lj7p6j8bPvtJKI8KQXJG2pI9pc1Z90rodrmmrOQRS_SnwVXxT2bOfOVDkNmI5QfB1XO47JrBnCLEKLzn-F9ct0Wb0QgtoJqYkZI32LWyJLin_RUFqIqKfNC_QSmJKimrFn_RbQM4sgURXLEzhycd&eid=8j7g4x-igj9clvr-4o


This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=8j7g4x-igj9clvr-4o&a=customerServiceUrl&ek=invite_guest

&copy; 2015 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

**Comment by stephenmsnow** [comment]
Stephen Snow would like to connect on LinkedIn. How would you like to respond?

Accept: https://www.linkedin.com/e/v2?e=8j7g4x-igttxzlc-1a&a=preRegInvite&ek=first_guest_reminder_01&li=13&m=hero&ts=accept_text&sharedKey=qBOkYmfz&invitationID=6067262158750699520

View Stephen Snow&#39;s profile: https://www.linkedin.com/comm/profile/view?id=AAsAABtG5F8B7_-uOfC80sK7Hdvirz3ZPd_nCbY&trk=eml-first_guest_reminder_01-hero-14-profile_text&trkEmail=eml-first_guest_reminder_01-hero-14-profile_text-null-8j7g4x%7Eigttxzlc%7E1a





You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=8j7g4x-igttxzlc-1a&t=lun&midToken=AQFmrFVhGEd36Q&ek=first_guest_reminder_01&li=16&m=unsub&ts=HTML&eid=8j7g4x-igttxzlc-1a&loid=AQGyduEoPnlQ9wAAAVDz2W0TA-SPe9cH2LB9VpKSjdK-iLqHa9TS3r22CwB_tCKe3ZCqfPCr6_eIhABxNumQaPUs70y6DrZmW1qzg4SXbUv-7PsRtiYDRzBYPc2_jelyWHZkWd7TaRA5Per1SQJI3MI22000KJnl6-MwDMb6NYPVq7NbjqREMpl6MDW_IVWRaywC

This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=8j7g4x-igttxzlc-1a&a=customerServiceUrl&ek=first_guest_reminder_01

&copy; 2015 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

**Comment by stephenmsnow** [comment]
Stephen Snow would like to connect on LinkedIn. How would you like to respond?

Accept: https://www.linkedin.com/e/v2?e=8j7g4x-ih3twuae-6d&a=preRegInvite&ek=second_guest_reminder_01&li=12&m=hero&ts=accept_text&sharedKey=qBOkYmfz&invitationID=6067262158750699520

View Stephen Snow&#39;s profile: https://www.linkedin.com/comm/profile/view?id=AAsAABtG5F8B7_-uOfC80sK7Hdvirz3ZPd_nCbY&trk=eml-second_guest_reminder_01-hero-13-profile_text&trkEmail=eml-second_guest_reminder_01-hero-13-profile_text-null-8j7g4x%7Eih3twuae%7E6d





You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=8j7g4x-ih3twuae-6d&t=lun&midToken=AQFmrFVhGEd36Q&ek=second_guest_reminder_01&li=15&m=unsub&ts=HTML&eid=8j7g4x-ih3twuae-6d&loid=AQEUPhLuIBr9nwAAAVEX5yvvT0F9vSO9fpOb85ojtPYiTQE8hq-AS6NGc3y1ftouSEme8D8lWJ12FXrmrShsat1pmEdVOkIUj89KX6ii2H9jgR2Plq-8flkvVhuURNx3BOJwtVDrz0cEACtioy1f0vk5lNOv4rxSYyDgoup-UEDwfZUdq75ggyq4rB46OVGR9-kr

This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=8j7g4x-ih3twuae-6d&a=customerServiceUrl&ek=second_guest_reminder_01

&copy; 2015 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

**Comment by rpavlik** [comment]
@stephenmsnow Stephen, this is the second time that our repositories have gotten spammed with LinkedIn invites. While LinkedIn shouldn't be sending email to reply@reply.github.com (and really, I disagree with their "email everyone in your address book" functionality entirely) and I've send in a support request to that effect, you'd be wise to not use that "spam everyone" functionality - it's poor security (sharing your email account credentials) and when things like this happen it might reflect poorly on you.

I'll be deleting that message and this one shortly, to clean up the project, but this was the only way I knew of to contact you.
